test/org/javarosa/core/model/Safe2014DagImplTest.java

```diff
@@ -0,0 +1,104 @@
+package org.javarosa.core.model;
+
+import org.javarosa.core.model.instance.FormInstance;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.debug.Event;
+import org.javarosa.debug.EventNotifier;
+import org.junit.Test;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.javarosa.test.utils.FormParserHelper.parse;
+import static org.javarosa.test.utils.ResourcePathHelper.r;
+import static org.junit.Assert.assertThat;
+
+public class Safe2014DagImplTest {
+
+    private final List<Event> dagEvents = new ArrayList<>();
+
+    private final EventNotifier eventNotifier = new EventNotifier() {
+
+        @Override
+        public void publishEvent(Event event) {
+            dagEvents.add(event);
+        }
+    };
+
+    @Test
+    public void deleteRepeatGroup_evaluatesTriggerables_dependedOnAllRepeatGroupSiblings() throws Exception {
+        // Given
+        final FormDef formDef =
+                parse(r("org/javarosa/core/model/repeat-group-with-position-calculation.xml")).formDef;
+
+        formDef.setEventNotifier(eventNotifier);
+        IDag dag = getIDagImplUnderTest(formDef);
+        assertThat(dag, instanceOf(Safe2014DagImpl.class));
+
+        final FormInstance mainInstance = formDef.getMainInstance();
+        final TreeElement parentElement = mainInstance.getRoot();
+
+        /*
+           We test a scenario when a second repeat group is deleted. There is no actual deletion that would've been
+           normally triggered by FormDef#deleteRepeat(FormIndex). Instead, a form instance with 4 repeat groups is used
+           which simulates situation when there were 5 repeats and the second one was deleted.
+           The first repeat has already a populated value  - '1' - which normally would've been calculated by DAG.
+           Rest of the repeat groups in the main instance have no values so we can test that those values are actually
+           calculated after dag.deleteRepeatGroup is called.
+
+           At position 0 there is a template TreeElement (multiplicity == -2)
+           At position 1 is a repeat group with an already populated populated
+           At positions 2 - 4 there are repeats with values supposed to be populated after deleteRepeatGroup is called
+        */
+        final TreeElement deletedElement = mainInstance.getRoot().getChildAt(2);
+        final TreeReference dummyDeletedRef = null;
+
+        // When
+        dag.deleteRepeatGroup(mainInstance, formDef.getEvaluationContext(), dummyDeletedRef, parentElement, deletedElement);
```

Comment:
I’m wondering if it would be safer, more realistic, and/or simpler to call formdef.deleteRepeat(FormIndex index) rather than getting the dag and calling deleteRepeatGroup directly.

Reply:
@dcbriccetti Well, I mostly wanted to test dag in isolation but it should be possible to test it via `formdef.deleteRepeat(FormIndex index)`. 
Since I'd like to avoid using API classes like `FormEntryModel` and/or `FormEntryController`, the `FormIndex index` would need to be constructed from scratch. After my time spent on #218 (where I was analyzing FormIndex inc/dec codebase) , it shouldn't be a problem now.

Reply:
@dcbriccetti I followed your suggestion and now the test looks way better. Thanks!